File: src/node/addons/spdlog.rs
use std::sync::Arc;
use spdlog::{critical, debug, error, info, sink::{FileSink, Sink}, trace, warn, Logger};
use tao::event_loop::EventLoopProxy;
use tokio::runtime::Runtime;
use wry::RequestAsyncResponder;
use crate::{backend::Backend, common::{respond_404, respond_ok}, node::node::AppEnv, types::ElectricoEvents};
use super::types::SPDLogCommand;

pub fn process_spdlog_command(_tokio_runtime:&Runtime, _app_env:&AppEnv,
    _proxy:EventLoopProxy<ElectricoEvents>,
    backend:&mut Backend,
    command:SPDLogCommand,
    responder:RequestAsyncResponder,
    _data_blob:Option<Vec<u8>>)  {
    
    match command {
        SPDLogCommand::CreateLogger { id, name, filepath } => {
            match FileSink::builder().path(filepath).build() {
                Ok(sink) => {
                    match Logger::builder().name(name).sink(Arc::new(sink)).build() {
                        Ok(logger) => {
                            backend.addon_state_insert(&id, logger);
                            respond_ok(responder);
                        },
                        Err(e) => {
                            log::error!("CreateLogger Error: {}", e);
                            respond_404(responder);
                        }
                    }
                },
                Err(e) => {
                    log::error!("CreateLogger Build FileSink Error: {}", e);
                    respond_404(responder);
                }
            }
        },
        SPDLogCommand::SetLogLevel { id, level } => {
            let logger:Option<&mut Logger> = backend.addon_state_get_mut(&id);
            if let Some(logger) = logger {
                match level {
                    super::types::SPDLogLevel::Trace => {
                        logger.set_level_filter(spdlog::LevelFilter::All);
                    },
                    crate::node::addons::types::SPDLogLevel::Debug => {
                        logger.set_level_filter(spdlog::LevelFilter::MoreSevereEqual(spdlog::Level::Debug));
                    },
                    crate::node::addons::types::SPDLogLevel::Error => {
                        logger.set_level_filter(spdlog::LevelFilter::MoreSevereEqual(spdlog::Level::Error));
                    },
                    crate::node::addons::types::SPDLogLevel::Info => {
                        logger.set_level_filter(spdlog::LevelFilter::MoreSevereEqual(spdlog::Level::Info));
                    },
                    crate::node::addons::types::SPDLogLevel::Warn => {
                        logger.set_level_filter(spdlog::LevelFilter::MoreSevereEqual(spdlog::Level::Warn));
                    },
                    crate::node::addons::types::SPDLogLevel::Critical => {
                        logger.set_level_filter(spdlog::LevelFilter::MoreSevereEqual(spdlog::Level::Critical));
                    },
                    crate::node::addons::types::SPDLogLevel::Off => {
                        logger.set_level_filter(spdlog::LevelFilter::Off);
                    }
                }
                respond_ok(responder);
            } else {
                log::error!("SPDLogger not found: {}", id);
                respond_404(responder);
            }
        },
        SPDLogCommand::Log { id, level, message } => {   
            let logger:Option<&mut Logger> = backend.addon_state_get_mut(&id);
            if let Some(logger) = logger {
                match level {
                    super::types::SPDLogLevel::Trace => {
                        trace!(logger:logger, "{}", message);
                    },
                    crate::node::addons::types::SPDLogLevel::Debug => {
                        debug!(logger:logger, "{}", message);
                    },
                    crate::node::addons::types::SPDLogLevel::Error => {
                        error!(logger:logger, "{}", message);
                    },
                    crate::node::addons::types::SPDLogLevel::Info => {
                        info!(logger:logger, "{}", message);
                    },
                    crate::node::addons::types::SPDLogLevel::Warn => {
                        warn!(logger:logger, "{}", message);
                    },
                    crate::node::addons::types::SPDLogLevel::Critical => {
                        critical!(logger:logger, "{}", message);
                    },
                    crate::node::addons::types::SPDLogLevel::Off => {
                        
                    }
                }
                logger.flush();
                respond_ok(responder);
            } else {
                log::error!("SPDLogger not found: {}", id);
                respond_404(responder);
            }
        },
        SPDLogCommand::Flush { id } => {
            let logger:Option<&mut Logger> = backend.addon_state_get_mut(&id);
            if let Some(logger) = logger {
                logger.flush();
                respond_ok(responder);
            } else {
                log::error!("SPDLogger not found: {}", id);
                respond_404(responder);
            }
        }
    }
}

File: src/js/backend/addons/spdlog.js
(function() {
    let uuidv4 = window.__uuidv4;
    let levels = ["Trace", "Debug", "Info", "Warn", "Error", "Critical", "Off"];
    let SPDLog = {
        setFlushOn: (level) => {
        },
        Logger: class {
            constructor(loggerType, name, filepath, maxFileSize, maxFiles) {
                this.id = uuidv4();
                this.name = name;
                this.filepath = filepath;
                $e_node.syncAddon_SPDLog_CreateLogger({"id":this.id, "name":name, "filepath":filepath});
                this.log = (function(level, message) {
                    $e_node.asyncAddon_SPDLog_Log({"id":this.id, "level":level, "message":message});
                }).bind(this);
                this.flush = (function() {
                    $e_node.asyncAddon_SPDLog_Flush({"id":this.id});
                }).bind(this);
                this.setLevel = (function(level) {
                    $e_node.asyncAddon_SPDLog_SetLogLevel({"id":this.id, "level":levels[level]});
                }).bind(this);
                this.setPattern = (function(pattern) {
                    //TODO
                }).bind(this);
                this.trace = (function(message) {
                    this.log("Trace", message);
                }).bind(this);
                this.debug = (function(message) {
                    this.log("Debug", message);
                }).bind(this);
                this.info = (function(message) {
                    this.log("Info", message);
                }).bind(this);
                this.warn = (function(message) {
                    this.log("Warn", message);
                }).bind(this);
                this.error = (function(message) {
                    this.log("Error", message);
                }).bind(this);
                this.clearFormatters=(function() {
                    
                }).bind(this);
            }
        }
    }
    window.__electrico.libs["spdlog"] = SPDLog;
})();